Relax ordering requirements for flags/commands to improve upon cli ergonomics

## Checklist

* [X ] Are you running the latest v2 release? The list of releases is [here](https://github.com/urfave/cli/releases).
* [X ] Did you check the manual for your release? The v2 manual is [here](https://github.com/urfave/cli/blob/master/docs/v2/manual.md)
* [ X] Did you perform a search about this feature? Here's the [Github guide](https://help.github.com/en/github/managing-your-work-on-github/using-search-to-filter-issues-and-pull-requests) about searching.

## What problem does this solve?

CLI ergonomics when editing/adding flags to command are very bad.

Quick example, consider how many times you have to move cursor back in commandline to go from

    $ cmd query users

to

    $ cmd --out=csv query --filter "project=kittens" users --with-permission=Deploy

vs

    $ cmd query users --with-permission=Deploy  --filter  "project=kittens", --out=csv 

It also "reads" worse:
 "output CSV on query project kittens on users with permissions to deploy:
 vs
 " query users that can deploy in project kittens, output csv"

Without forced ordering you can add a flag at the end at any point so ad-hoc usage of commands is much more streamlined, it also allows writing CLI scripts with arguments grouped by what makes more sense, not by how program author structured it.

Now the long winded example, consider the following:
* App has global flags applicable to all subcommands like
  * change output type (for example switch between machine-parseable and user-frien
  * server address
  * config/credentials pass
  * environment
   * etc
* Application has subcommand with their own flags user might want to iterate
   
Let's say app has a subcommand to query systems state

    cmd query

Which has `--filter` string option, and this command have subcommands for various parts of system

    cmd query nodes
    cmd query services
    cmd query users
    
with various sub-subcommand specific switches.
Let's say user wants to only see stuff about a certain project. No worries, we have `--filter` flag for it

    $ cmd query --filter project=kittens

User now wants to see who has access to that project. But we have options just for that

    $ cmd query --filter project=kittens users 

So far so good. Let's see which of them can deploy the application

    $ cmd query --filter project=kittens users --with-permission=Deploy

Okay, now let's compare it with another project...

    $ cmd query --filter project=dogs users --with-permission=Deploy

Now we need to go back 4 words (4x `C-b`) and we can start changing it. But we have few more projects to check? How about just move filter at the end ?

    $ cmd query  users --with-permission=Deploy --filter project=dogs
    flag provided but not defined: -filter

Okay, we can't. Bit annoying but it is only few commands. 

    $ for project in kittens dogs otters ducks ; do cmd query --filter "project=$project" users --with-permission=Deploy ; done

Right, it does what we wanted. Now onto writing that report. There is a `csv` output format so it should be trivial.
 
    $ for project in kittens dogs otters ducks ; do cmd query --filter "project=$project" users --with-permission=Deploy --out=csv; done
    flag provided but not defined: -out

Oh, right, *go C-b NINE times* because that needs to be after a command but before subcommand.

    $ for project in kittens dogs otters ducks ; do cmd --out=csv query --filter "project=$project" users --with-permission=Deploy ; done

Want to bump it to verbose or enable debug ? Go back nearly to the start of the line. Want to set environment or target server ? Same. Only real option to have convenient ability to change the parameters used in whole app is duplicating them as command-local variables.

The migration manual mentions something that "it is more POSIX-like" but POSIX conventions never had a notion of subcommands in the first place, the "stuff after options" was not a subcommand but also command's parameters (usually file to operate on), and even now most of the say `coreutils` tools allow for having options in any place just for convenience (GNU's `ls -la /tmp` and `ls /tmt -la` have same effect for example)


## Solution description

Every `--option` in commandline should be considered for a flag, from closest "leaf" subcommand to the root

so

    $ cmd --env=dev query --filter type=sqldb
    $ cmd query --filter type=sqldb --env=dev

should be equivalent. Sometimes even having options before subcommand makes sense, like

    $ cmd --filter="type=sqldb" query users
    $ cmd --filter="type=sqldb" query nodes
    $ cmd --filter="type=sqldb" query services

## Describe alternatives you've considered

Copying every flag to local "works" but it isn't exactly elegant, and messes up with generated help.

In principle it should be as easy as replacing `"flag"` import with https://github.com/spf13/pflag but it seems not synchronized with upstreams and we have few errors:

```
./context.go:111:39: undefined: pflag.Getter
./flag_float64_slice.go:140:12: cannot use f.Value (type *Float64Slice) as type pflag.Value in argument to set.Var:
	*Float64Slice does not implement pflag.Value (missing Type method)
./flag_float64_slice.go:158:26: impossible type assertion:
	*Float64Slice does not implement pflag.Value (missing Type method)
./flag_generic.go:83:12: cannot use f.Value (type Generic) as type pflag.Value in argument to set.Var:
	Generic does not implement pflag.Value (missing Type method)
./flag_int64_slice.go:139:12: cannot use f.Value (type *Int64Slice) as type pflag.Value in argument to set.Var:
	*Int64Slice does not implement pflag.Value (missing Type method)
./flag_int64_slice.go:154:26: impossible type assertion:
	*Int64Slice does not implement pflag.Value (missing Type method)
./flag_int_slice.go:150:12: cannot use f.Value (type *IntSlice) as type pflag.Value in argument to set.Var:
	*IntSlice does not implement pflag.Value (missing Type method)
./flag_int_slice.go:168:26: impossible type assertion:
	*IntSlice does not implement pflag.Value (missing Type method)
./flag_string_slice.go:151:13: cannot use f.Destination (type *StringSlice) as type pflag.Value in argument to set.Var:
	*StringSlice does not implement pflag.Value (missing Type method)
./flag_string_slice.go:155:12: cannot use f.Value (type *StringSlice) as type pflag.Value in argument to set.Var:
	*StringSlice does not implement pflag.Value (missing Type method)
./flag_string_slice.go:155:12: too many errors
```

--- a/app_test.go
+++ b/app_test.go
@@ -1461,6 +1461,10 @@ func TestApp_AfterFunc(t *testing.T) {
 		reset
 	*/
 	counts = &opCounts{}
+	// reset the flags since they are set previously
+	app.Flags = []Flag{
+		&StringFlag{Name: "opt"},
+	}
 
 	// run with none args
 	err = app.Run([]string{"command"})
@@ -3052,4 +3056,132 @@ func TestFlagAction(t *testing.T) {
 			}
 		})
 	}
+}
+
+func TestPersistentFlag(t *testing.T) {
+
+	var topInt, topPersistentInt, subCommandInt, appOverrideInt int
+	var appFlag string
+	var appOverrideCmdInt int64
+	var appSliceFloat64 []float64
+	var persistentAppSliceInt []int64
+
+	a := &App{
+		Flags: []Flag{
+			&StringFlag{
+				Name:        "persistentAppFlag",
+				Persistent:  true,
+				Destination: &appFlag,
+			},
+			&Int64SliceFlag{
+				Name:        "persistentAppSliceFlag",
+				Persistent:  true,
+				Destination: &persistentAppSliceInt,
+			},
+			&Float64SliceFlag{
+				Name:       "persistentAppFloatSliceFlag",
+				Persistent: true,
+				Value:      []float64{11.3, 12.5},
+			},
+			&IntFlag{
+				Name:        "persistentAppOverrideFlag",
+				Persistent:  true,
+				Destination: &appOverrideInt,
+			},
+		},
+		Commands: []*Command{
+			{
+				Name: "cmd",
+				Flags: []Flag{
+					&IntFlag{
+						Name:        "cmdFlag",
+						Destination: &topInt,
+					},
+					&IntFlag{
+						Name:        "cmdPersistentFlag",
+						Persistent:  true,
+						Destination: &topPersistentInt,
+					},
+					&Int64Flag{
+						Name:        "paof",
+						Aliases:     []string{"persistentAppOverrideFlag"},
+						Destination: &appOverrideCmdInt,
+					},
+				},
+				Subcommands: []*Command{
+					{
+						Name: "subcmd",
+						Flags: []Flag{
+							&IntFlag{
+								Name:        "cmdFlag",
+								Destination: &subCommandInt,
+							},
+						},
+						Action: func(ctx *Context) error {
+							appSliceFloat64 = ctx.Float64Slice("persistentAppFloatSliceFlag")
+							return nil
+						},
+					},
+				},
+			},
+		},
+	}
+
+	err := a.Run([]string{"app",
+		"--persistentAppFlag", "hello",
+		"--persistentAppSliceFlag", "100",
+		"--persistentAppOverrideFlag", "102",
+		"cmd",
+		"--cmdFlag", "12",
+		"--persistentAppSliceFlag", "102",
+		"--persistentAppFloatSliceFlag", "102.455",
+		"--paof", "105",
+		"subcmd",
+		"--cmdPersistentFlag", "20",
+		"--cmdFlag", "11",
+		"--persistentAppFlag", "bar",
+		"--persistentAppSliceFlag", "130",
+		"--persistentAppFloatSliceFlag", "3.1445",
+	})
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if appFlag != "bar" {
+		t.Errorf("Expected 'bar' got %s", appFlag)
+	}
+
+	if topInt != 12 {
+		t.Errorf("Expected 12 got %d", topInt)
+	}
+
+	if topPersistentInt != 20 {
+		t.Errorf("Expected 20 got %d", topPersistentInt)
+	}
+
+	// this should be changed from app since
+	// cmd overrides it
+	if appOverrideInt != 102 {
+		t.Errorf("Expected 102 got %d", appOverrideInt)
+	}
+
+	if subCommandInt != 11 {
+		t.Errorf("Expected 11 got %d", subCommandInt)
+	}
+
+	if appOverrideCmdInt != 105 {
+		t.Errorf("Expected 105 got %d", appOverrideCmdInt)
+	}
+
+	expectedInt := []int64{100, 102, 130}
+	if !reflect.DeepEqual(persistentAppSliceInt, expectedInt) {
+		t.Errorf("Expected %v got %d", expectedInt, persistentAppSliceInt)
+	}
+
+	expectedFloat := []float64{102.455, 3.1445}
+	if !reflect.DeepEqual(appSliceFloat64, expectedFloat) {
+		t.Errorf("Expected %f got %f", expectedFloat, appSliceFloat64)
+	}
+
 }

--- a/command.go
+++ b/command.go
@@ -146,7 +146,7 @@ func (c *Command) Run(cCtx *Context, arguments ...string) (err error) {
 	}
 
 	a := args(arguments)
-	set, err := c.parseFlags(&a, cCtx.shellComplete)
+	set, err := c.parseFlags(&a, cCtx)
 	cCtx.flagSet = set
 
 	if c.isRoot {
@@ -308,7 +308,7 @@ func (c *Command) suggestFlagFromError(err error, command string) (string, error
 	return fmt.Sprintf(SuggestDidYouMeanTemplate, suggestion) + "\n\n", nil
 }
 
-func (c *Command) parseFlags(args Args, shellComplete bool) (*flag.FlagSet, error) {
+func (c *Command) parseFlags(args Args, ctx *Context) (*flag.FlagSet, error) {
 	set, err := c.newFlagSet()
 	if err != nil {
 		return nil, err
@@ -318,13 +318,42 @@ func (c *Command) parseFlags(args Args, shellComplete bool) (*flag.FlagSet, erro
 		return set, set.Parse(append([]string{"--"}, args.Tail()...))
 	}
 
-	err = parseIter(set, c, args.Tail(), shellComplete)
-	if err != nil {
+	for pCtx := ctx.parentContext; pCtx != nil; pCtx = pCtx.parentContext {
+		if pCtx.Command == nil {
+			continue
+		}
+
+		for _, fl := range pCtx.Command.Flags {
+			pfl, ok := fl.(PersistentFlag)
+			if !ok || !pfl.IsPersistent() {
+				continue
+			}
+
+			applyPersistentFlag := true
+			set.VisitAll(func(f *flag.Flag) {
+				for _, name := range fl.Names() {
+					if name == f.Name {
+						applyPersistentFlag = false
+						break
+					}
+				}
+			})
+
+			if !applyPersistentFlag {
+				continue
+			}
+
+			if err := fl.Apply(set); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if err := parseIter(set, c, args.Tail(), ctx.shellComplete); err != nil {
 		return nil, err
 	}
 
-	err = normalizeFlags(c.Flags, set)
-	if err != nil {
+	if err := normalizeFlags(c.Flags, set); err != nil {
 		return nil, err
 	}
 

--- a/flag.go
+++ b/flag.go
@@ -172,6 +172,12 @@ type CategorizableFlag interface {
 	GetCategory() string
 }
 
+// PersistentFlag is an interface to enable detection of flags which are persistent
+// through subcommands
+type PersistentFlag interface {
+	IsPersistent() bool
+}
+
 func flagSet(name string, flags []Flag) (*flag.FlagSet, error) {
 	set := flag.NewFlagSet(name, flag.ContinueOnError)
 

--- a/flag_impl.go
+++ b/flag_impl.go
@@ -41,8 +41,9 @@ type FlagBase[T any, C any, VC ValueCreator[T, C]] struct {
 	FilePath    string // file path to load value from
 	Usage       string // usage string for help output
 
-	Required bool // whether the flag is required or not
-	Hidden   bool // whether to hide the flag in help output
+	Required   bool // whether the flag is required or not
+	Hidden     bool // whether to hide the flag in help output
+	Persistent bool // whether the flag needs to be applied to subcommands as well
 
 	Value       T  // default value for this flag if not set by from any source
 	Destination *T // destination pointer for value when set
@@ -58,6 +59,7 @@ type FlagBase[T any, C any, VC ValueCreator[T, C]] struct {
 
 	// unexported fields for internal use
 	hasBeenSet bool  // whether the flag has been set from env or file
+	applied    bool  // whether the flag has been applied to a flag set already
 	creator    VC    // value creator for this flag type
 	value      Value // value representing this flag's value
 }
@@ -73,35 +75,41 @@ func (f *FlagBase[T, C, V]) GetValue() string {
 
 // Apply populates the flag given the flag set and environment
 func (f *FlagBase[T, C, V]) Apply(set *flag.FlagSet) error {
-	newVal := f.Value
+	// if flag has been applied then it wouldnt have been set from
+	// env or file as well.
+	// TODO move this phase into a separate flag initialization function
+	if !f.applied {
+		newVal := f.Value
 
-	if val, source, found := flagFromEnvOrFile(f.EnvVars, f.FilePath); found {
-		tmpVal := f.creator.Create(f.Value, new(T), f.Config)
-		if val != "" || reflect.TypeOf(f.Value).Kind() == reflect.String {
-			if err := tmpVal.Set(val); err != nil {
-				return fmt.Errorf("could not parse %q as %T value from %s for flag %s: %s", val, f.Value, source, f.Name, err)
+		if val, source, found := flagFromEnvOrFile(f.EnvVars, f.FilePath); found {
+			tmpVal := f.creator.Create(f.Value, new(T), f.Config)
+			if val != "" || reflect.TypeOf(f.Value).Kind() == reflect.String {
+				if err := tmpVal.Set(val); err != nil {
+					return fmt.Errorf("could not parse %q as %T value from %s for flag %s: %s", val, f.Value, source, f.Name, err)
+				}
+			} else if val == "" && reflect.TypeOf(f.Value).Kind() == reflect.Bool {
+				val = "false"
+				if err := tmpVal.Set(val); err != nil {
+					return fmt.Errorf("could not parse %q as %T value from %s for flag %s: %s", val, f.Value, source, f.Name, err)
+				}
 			}
-		} else if val == "" && reflect.TypeOf(f.Value).Kind() == reflect.Bool {
-			val = "false"
-			if err := tmpVal.Set(val); err != nil {
-				return fmt.Errorf("could not parse %q as %T value from %s for flag %s: %s", val, f.Value, source, f.Name, err)
-			}
+
+			newVal = tmpVal.Get().(T)
+			f.hasBeenSet = true
 		}
 
-		newVal = tmpVal.Get().(T)
-		f.hasBeenSet = true
-	}
-
-	if f.Destination == nil {
-		f.value = f.creator.Create(newVal, new(T), f.Config)
-	} else {
-		f.value = f.creator.Create(newVal, f.Destination, f.Config)
+		if f.Destination == nil {
+			f.value = f.creator.Create(newVal, new(T), f.Config)
+		} else {
+			f.value = f.creator.Create(newVal, f.Destination, f.Config)
+		}
 	}
 
 	for _, name := range f.Names() {
 		set.Var(f.value, name, f.Usage)
 	}
 
+	f.applied = true
 	return nil
 }
 
@@ -178,7 +186,13 @@ func (f *FlagBase[T, C, V]) RunAction(ctx *Context) error {
 	return nil
 }
 
+// IsSliceFlag returns true if the value type T is of kind slice
 func (f *FlagBase[T, C, VC]) IsSliceFlag() bool {
 	// TBD how to specify
 	return reflect.TypeOf(f.Value).Kind() == reflect.Slice
+}
+
+// IsPersistent returns true if flag needs to be persistent across subcommands
+func (f *FlagBase[T, C, VC]) IsPersistent() bool {
+	return f.Persistent
 }

--- a/flag_impl.go
+++ b/flag_impl.go
@@ -75,10 +75,13 @@ func (f *FlagBase[T, C, V]) GetValue() string {
 
 // Apply populates the flag given the flag set and environment
 func (f *FlagBase[T, C, V]) Apply(set *flag.FlagSet) error {
-	// if flag has been applied then it wouldnt have been set from
-	// env or file as well.
 	// TODO move this phase into a separate flag initialization function
-	if !f.applied {
+	// if flag has been applied previously then it would have already been set
+	// from env or file. So no need to apply the env set again. However
+	// lots of units tests prior to persistent flags assumed that the
+	// flag can be applied to different flag sets multiple times while still
+	// keeping the env set.
+	if !f.applied || !f.Persistent {
 		newVal := f.Value
 
 		if val, source, found := flagFromEnvOrFile(f.EnvVars, f.FilePath); found {

--- a/godoc-current.txt
+++ b/godoc-current.txt
@@ -780,8 +780,9 @@ type FlagBase[T any, C any, VC ValueCreator[T, C]] struct {
 	FilePath    string // file path to load value from
 	Usage       string // usage string for help output
 
-	Required bool // whether the flag is required or not
-	Hidden   bool // whether to hide the flag in help output
+	Required   bool // whether the flag is required or not
+	Hidden     bool // whether to hide the flag in help output
+	Persistent bool // whether the flag needs to be applied to subcommands as well
 
 	Value       T  // default value for this flag if not set by from any source
 	Destination *T // destination pointer for value when set
@@ -826,6 +827,9 @@ func (f *FlagBase[T, C, V]) GetValue() string
     GetValue returns the flags value as string representation and an empty
     string if the flag takes no value at all.
 
+func (f *FlagBase[T, C, VC]) IsPersistent() bool
+    IsPersistent returns true if flag needs to be persistent across subcommands
+
 func (f *FlagBase[T, C, V]) IsRequired() bool
     IsRequired returns whether or not the flag is required
 
@@ -833,6 +837,7 @@ func (f *FlagBase[T, C, V]) IsSet() bool
     IsSet returns whether or not the flag has been set through env or file
 
 func (f *FlagBase[T, C, VC]) IsSliceFlag() bool
+    IsSliceFlag returns true if the value type T is of kind slice
 
 func (f *FlagBase[T, C, V]) IsVisible() bool
     IsVisible returns true if the flag is not hidden, otherwise false
@@ -939,6 +944,12 @@ type OnUsageErrorFunc func(cCtx *Context, err error, isSubcommand bool) error
     displaying customized usage error messages. This function is able to replace
     the original error messages. If this function is not set, the "Incorrect
     usage" is displayed and the execution is interrupted.
+
+type PersistentFlag interface {
+	IsPersistent() bool
+}
+    PersistentFlag is an interface to enable detection of flags which are
+    persistent through subcommands
 
 type RequiredFlag interface {
 	Flag

--- a/testdata/godoc-v3.x.txt
+++ b/testdata/godoc-v3.x.txt
@@ -780,8 +780,9 @@ type FlagBase[T any, C any, VC ValueCreator[T, C]] struct {
 	FilePath    string // file path to load value from
 	Usage       string // usage string for help output
 
-	Required bool // whether the flag is required or not
-	Hidden   bool // whether to hide the flag in help output
+	Required   bool // whether the flag is required or not
+	Hidden     bool // whether to hide the flag in help output
+	Persistent bool // whether the flag needs to be applied to subcommands as well
 
 	Value       T  // default value for this flag if not set by from any source
 	Destination *T // destination pointer for value when set
@@ -826,6 +827,9 @@ func (f *FlagBase[T, C, V]) GetValue() string
     GetValue returns the flags value as string representation and an empty
     string if the flag takes no value at all.
 
+func (f *FlagBase[T, C, VC]) IsPersistent() bool
+    IsPersistent returns true if flag needs to be persistent across subcommands
+
 func (f *FlagBase[T, C, V]) IsRequired() bool
     IsRequired returns whether or not the flag is required
 
@@ -833,6 +837,7 @@ func (f *FlagBase[T, C, V]) IsSet() bool
     IsSet returns whether or not the flag has been set through env or file
 
 func (f *FlagBase[T, C, VC]) IsSliceFlag() bool
+    IsSliceFlag returns true if the value type T is of kind slice
 
 func (f *FlagBase[T, C, V]) IsVisible() bool
     IsVisible returns true if the flag is not hidden, otherwise false
@@ -939,6 +944,12 @@ type OnUsageErrorFunc func(cCtx *Context, err error, isSubcommand bool) error
     displaying customized usage error messages. This function is able to replace
     the original error messages. If this function is not set, the "Incorrect
     usage" is displayed and the execution is interrupted.
+
+type PersistentFlag interface {
+	IsPersistent() bool
+}
+    PersistentFlag is an interface to enable detection of flags which are
+    persistent through subcommands
 
 type RequiredFlag interface {
 	Flag

